File: aoc15/day_16/day_16.ml
let parse_thing t =
  let re = Str.regexp {|\(.* \|^\)\([a-z]+\): \([0-9]+\)$|} in
  let _ = Str.string_match re t 0 in
  (Str.matched_group 2 t, int_of_string (Str.matched_group 3 t))
in
let parse_sue line = List.map parse_thing (String.split_on_char ',' line) in
let sues = Input.parse_lines parse_sue "input" in

let tape = Input.parse_lines parse_thing "tape" in

let match_exact tape thing = List.exists (fun t -> t = thing) tape in
let match_range tape thing =
  let match_range_things (n1, v1) (n2, v2) =
    n1 = n2
    &&
    match n1 with
    | "cats" | "trees" -> v1 > v2
    | "pomeranians" | "goldfish" -> v1 < v2
    | _ -> v1 = v2
  in
  List.exists (match_range_things thing) tape
in

let first = Utils.find_index (List.for_all (match_exact tape)) sues in
let second = Utils.find_index (List.for_all (match_range tape)) sues in
Printf.printf "First: %d\n" (first + 1);
Printf.printf "Second: %d\n" (second + 1)

File: aoc15/go/day_16/day_16.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Sue map[string]int

func parse(line string) Sue {
	str := line[strings.IndexByte(line, ':')+2:]
	things := strings.Split(str, ", ")
	sue := make(map[string]int)
	for _, thing := range things {
		ts := strings.Split(thing, ": ")
		n, _ := strconv.Atoi(ts[1])
		sue[ts[0]] = n
	}
	return sue
}

func firstCheck(sue, sender Sue) bool {
	for t, n := range sue {
		if sender[t] != n {
			return false
		}
	}
	return true
}

func secondCheck(sue, sender Sue) bool {
	for t, n := range sue {
		if t == "cats" || t == "trees" {
			if n <= sender[t] {
				return false
			}
		} else if t == "pomeranians" || t == "goldfish" {
			if n >= sender[t] {
				return false
			}
		} else if sender[t] != n {
			return false
		}
	}
	return true
}

func main() {
	input, _ := os.Open("input")
	scanner := bufio.NewScanner(input)

	sender := Sue{
		"children":    3,
		"cats":        7,
		"samoyeds":    2,
		"pomeranians": 3,
		"akitas":      0,
		"vizslas":     0,
		"goldfish":    5,
		"trees":       3,
		"cars":        2,
		"perfumes":    1,
	}

	i := 1
	for scanner.Scan() {
		sue := parse(scanner.Text())
		if firstCheck(sue, sender) {
			fmt.Println("First:", i)
		}
		if secondCheck(sue, sender) {
			fmt.Println("Second:", i)
		}
		i++
	}
}
